### ISSUE TITLE
Postgresql - Error - No Driver has been registered with name

### ISSUE THREAD

**Issue by indrayanivaze** [aws/aws-secretsmanager-jdbc#44]
I am using aws secrets manager to connect to postgresql db hosted on aws. When I try to run locally with groovy grails app, everything works seamlessly without any errors. But when I try to deploy war on tomcat 8, I see the error : 

```
Caused by: java.lang.IllegalStateException: No Driver has been registered with name, org.postgresql.Driver. Please check your system properties or secretsmanager.properties for typos. Also ensure that the Driver registers itself.
	at com.amazonaws.secretsmanager.sql.AWSSecretsManagerDriver.getWrappedDriver(AWSSecretsManagerDriver.java:244)
	at com.amazonaws.secretsmanager.sql.AWSSecretsManagerDriver.acceptsURL(AWSSecretsManagerDriver.java:257)
	at com.zaxxer.hikari.util.DriverDataSource.<init>(DriverDataSource.java:83)
	at com.zaxxer.hikari.pool.PoolUtilities.initializeDataSource(PoolUtilities.java:114)
```

**Comment by daniel097541** [comment]
Did you solve this?

**Comment by indrayanivaze** [comment]
Nope, I haven't got a solution for it. One of the solutions that raghusujju gave was - "I overritten getDefaultDriverClass method inside AWSSecretsManagerPostgreSQLDriver.java ( as I was using postgres) to load the postgresql driver. JUst do a Class.forName("org.postgresql.Driver", false, this.getClass().getClassLoader());. AFter this the issue with driver load was gone.". But I have not tried it yet

**Comment by daniel097541** [comment]
I solved it, you will need to copy the whole class:  AWSSecretsManagerPostgreSQLDriver and add it into your code, this is because their class is final and cannot be extended, then modify the static brackets where they instanciate the driver and instanciate a new org.postgresql.Driver  just before, this is working for me:
https://pastebin.com/rTWGPYPU

**Comment by indrayanivaze** [comment]
ohh.. Thank you very much for letting me know that it worked ...In all did you create 2 classes ? CustomAWSSecretsManagerPostgreSQLDriver and CustomAWSSecretsManagerDriver ? Do you mind also adding imports ? I am seeing some issues with the imports

**Comment by shmam** [comment]
Encountered this same issue, and only solved it by editing the source code and instantiating a new instance of the default driver. Would like to see this resolved without having to edit the source code though. Is there any way to get this addressed by the devs?

**Comment by ehelbig1** [comment]
Can this change be added to the repo? Happy to create a PR if that will help expedite things.

**Comment by fridovanorden** [comment]
Would be a very welcome addition. If done properly in the AWSSecretsManagerDriver class all other supported databases benefit as well. Best place IMO is in the getWrappedDriver() method:

    protected void ensureRealDriverRegistered() {
    	try {
		// Load the real driver through the same class loader as the wrapper
		Class<Driver> cls = (Class<Driver>) getClass().getClassLoader().loadClass(this.realDriverClass);
		DriverManager.registerDriver(cls.getDeclaredConstructor().newInstance()); 
	} catch (Exception e) {
		e.printStackTrace();
	}
    }

    public Driver getWrappedDriver() {
    	ensureRealDriverRegistered();
         // existing code here

**Comment by realvictorprm** [comment]
And the problem still remains open :/

**Comment by devdeepak02** [comment]
This thread was a life saver.  thank you.

**Comment by willtong1234** [comment]
This looks like a duplicate of https://github.com/aws/aws-secretsmanager-jdbc/issues/26, but I'll keep it open until we can confirm it is indeed the same issue. We are looking into it.

**Comment by garg01puneet** [comment]
@indrayanivaze  can you please share if you are able to make workaround for this?
@fridovanorden are you suggesting to change the source code of this jar?

@willtong1234 , can you please update if any solution for this is available, or if any suggested workaround is available for this?

**Comment by jinspark-lab** [comment]
Same error with mysql Driver. Any workaround? It is so complicated to change all codes.

**Comment by lludsubaru** [comment]
I guess people still using Tomcat are too old-school? But I'm using it ( beanstalk ) and trying to use PostgreSQL. I added this code ( including setting initialization to true ) to my version to get it to work. I don't have ( easy ) access to the other databases so can't easily test if it works for anything else. If this is just specific to PostgreSQL maybe I should just add code in the PostgreSQL class. 

```
public Driver getWrappedDriver() {
      try {
        Class.forName( this.realDriverClass, true, this.getClass().getClassLoader() );
      } catch( ClassNotFoundException ex ) {
        throw new IllegalStateException( "Unable to pre-load (Class.forName) "+this.realDriverClass+" to force initialization" );
      }

```

**Comment by atul-chegg** [comment]
This is also issue for us. 
Our application war file is run via Jetty.

**Comment by lludsubaru** [comment]
I haven't does this before; but I'm wondering if there is a way to get this fixed? I'm not sure it's the best way, but I would probably make a method called loadDriver which would get called by the initialization routine and then change the target classes to not be final so someone could extend them and provide a loadDriver implementation if needed.  I'm happy to ( learn how and ) submit the changes but didn't want to bother if either A ) It wasn't a good design or B ) it would never be accepted. Can an review let me know if that makes any sense?

**Comment by kboxeth** [comment]
This should be resolved by #31 , feel free to reach out if issues continue to arise.

**Comment by hamzou247** [comment]
Hello, 

I am trying to fix this for a spring-boot app. Does anyone know to fix it in this context? Does the fix require editing aws-secretsmanager-jdbc code? I couldn't get it right :)
